### ISSUE TITLE
Train model for GM segmentation

### ISSUE THREAD

**Issue by Nilser3** [ivadomed/model-gm-contrast-region-agnostic#2]
# Data

### Multicentric dataset (for train/validation/testing)

- Number of 3D volumes : 1091
- Number of 2D slices : 14075

**Table 01**      
| Dataset               | Sequence         |Category         | Region           | In-plane res.    | N Subjects | Contrast | N of 2D slices |
|-------------------------|------------------|----------------|-------------------------|--------------------------|-------------|------------------------|------------------|
| marseille-7T-T2star    | 7T T2star        |  HC MS ALS        | cervical       | 0.18x0.18  0.22x0.22 | 75          |  GM_bright WM_dark CSF_bright                     | 753              | 
| marseille-7T-MP2RAGE     | 7T T1map       | HC MS ALS AMS    | cervical sup. and inf. (2 runs)    | 0.3x0.3           |     91      | GM_bright WM_dark_CSF_bright                     | 1791              | 
| marseille-t2s-template | 3T T2star      | HC               | cervical, torax, lumbar | 0.47x0.47               | 25          | GM_bright WM_dark CSF_bright                     | 346             | 
| inspired               | 3T T2star         | HC DCM SCI       | cervical           | 0.5×0.5         | 61          | GM_bright WM_dark CSF_bright                     | 579              | 
| gmseg-challenge-2016   | 3T T2star        |HC               | cervical             | 0.6×0.6       | 40          | GM_bright WM_dark CSF_bright                     | 469              | 
| lumbar-vanderbilt      | 3T T2star       | HC               | lumbar                 | 0.3×0.3           | 53        | GM_bright WM_dark CSF_bright                     | 727              | 
| sct-testing-large      | 3T T2star        | HC MS DCM   | cervical sup. and inf.  (2 runs)  | 0.5×0.5         | 333         |GM_bright WM_dark CSF_bright                    | 4272             | 
| sct-testing-large     | 3T MTon_MTR     | HC MS DCM   | cervical sup. and inf. (2 runs)  | 0.9×0.9       | 42          | GM_bright WM_dark CSF_bright                    | 522             | 
| dcm-brno               | 3T T1w ax        | HC               | cervical                   | 0.35×0.35             | 65          | GM_bright WM_dark CSF_dark                     | 2697             | 
| hc-ucsf-psir           | 3T PSIR ax         | HC               | c3                         | 0.8x0.8                 | 110         | GM_dark WM_bright CSF_dark                      | 110             | 
| marseille-7T-MP2RAGE     | 7T UNIT1   | HC MS ALS AMS    | cervical sup. and inf.  (2 runs)   | 0.3x0.3       |     91      | GM_dark WM_bright CSF_dark                     | 1791              | 

## Details of **Table 01** 

1. Multicentric datasets by sequence 
![barplot_datasets_vs_slices](https://github.com/user-attachments/assets/14d42b54-1651-4145-a2aa-640afff043e5)

2. Variety of contrast by sequences (contrast agnostic)
![barplot_contrasts_vs_slices](https://github.com/user-attachments/assets/07f7669d-ce46-47ed-92cb-3e45c3344f43)

3. Variety of acquired regions by category (region agnostic)
![barplot_region_vs_slices](https://github.com/user-attachments/assets/dd5c1c10-9309-4cf0-bbb7-0fe752b610d8)

**Comment by Nilser3** [comment]
### Dataset without GM manual segmentations (for testing)

| Dataset               | Sequence         |Category         | Region           | In-plane res. | Volumes |
|-------------------------|------------------|----------------|-------------------------|--------------------------|-------------|
| umass-ms-ge-pioneer3       |  3T PD  |    MS   |    Cervical      |     0.37×0.37     |     494      |    
| ms-karolinska-2020      | 3T T2star   | MS      |   Cervical       |    0.35×0.35       |       74    |   
| ms-basel-2020      | 3T T2star   |   MS    |    Cervical      |      0.35×0.35     |    36       |    
| vanderbilt-7t-swi      | 7T T2star, 7T SWI | HC, MS           | Cervical                   | 0.3×0.3               | 36         | 
| levin-stroke           | 3T T2star        | Stroke           | Cervical                   | 0.5×0.5               | 10         | 
| philadelphi-pediatric  | 3T T2star     | Pediatric        | Cervical                   | 0.5×0.5               | 24         | 
| umass-ms-siemens-espree1.5      | 1.5T PD   |   MS    |    Cervical      |   0.39×0.39       |     45      | 
| hc-leipzig-7t-mp2rage      | 7T UNI   |   HC    |     Cervical     |      0.7×0.7    |     25      |

**Comment by Nilser3** [comment]
# Experiment 1

## Preliminary results

### Training details:

1. Data : Table01, 80% for Train/validation and 20% Testing (by Dataset)
2. Testing dataset : 233 volumes (3123 2D slices)
3. Approach : 2D single-class (Binary GM)
4. Framework to train : nnUnetV2 
5. CV Folds : 5 folds
6. Epochs : 500

### Comparison of Dice Score Metric of 5 models in the Testing dataset (measured on 2D slices)

Stat using `Wilcoxon` test

![Image](https://github.com/user-attachments/assets/af6f2dd0-99ad-4d4c-a4d9-8e31ca397fce)

**Metrics description:**

| Statistic | dice_gm_fold0 | dice_gm_fold1 | dice_gm_fold2 | dice_gm_fold3 | dice_gm_fold4 |
|-----------|---------------|---------------|---------------|---------------|---------------|
| Count     | 3123.000000   | 3123.000000   | 3123.000000   | 3123.000000   | 3123.000000   |
| Mean      | 0.788741      | 0.796518      | 0.802884      | 0.801493      | 0.802258      |
| Std       | 0.208445      | 0.199479      | 0.188372      | 0.190161      | 0.189920      |
| Min       | 0.000000      | 0.000000      | 0.000000      | 0.000000      | 0.000000      |
| 25%       | 0.790786      | 0.792627      | 0.800000      | 0.797297      | 0.796801      |
| 50%       | 0.854369      | 0.857143      | 0.857955      | 0.860140      | 0.859813      |
| 75%       | 0.894573      | 0.895679      | 0.898131      | 0.896552      | 0.896327      |
| Max       | 0.980392      | 0.983051      | 0.989691      | 0.974093      | 0.983607      |


## Conclusion 
`Fold 2` is significantly better than the other models, so it is chosen as the `best model` and will be applied to further experiments.

**Comment by Nilser3** [comment]
Comparing the `Fold 2` (best model) with `sct_deepseg_gm`, by region and by contrast : 

![Image](https://github.com/user-attachments/assets/5ef9e428-ad54-4ad4-9155-a54043d91d07)

In general the contrast-agnostic model performs better than `sct_deepseg_gm`, but we need to compare it with the benchmark dedicated to each contrast (see issue #3 )

**Comment by Nilser3** [comment]
## Training on cropped images

To avoid the problem reported here: https://github.com/ivadomed/model-gm-contrast-region-agnostic/issues/3#issuecomment-2612358203 , the training database was cropped based on the GM mask (saving the entire XY dimension) and the `contrast-region-agostic` model was retrained (`contrast agnostic model trained on cropped images`) and comparing with previous model `contrast agnostic model trained on full FOV` , we can see that `contrast agnostic model trained on cropped images` performs much better on test subjects.

![Image](https://github.com/user-attachments/assets/d46c5651-86ce-4c46-a47f-133fe8fb2ab3)

Even comparing it with benchmarks:

| Contrast      | benchmarks| 
|------------------|----------------------------|
| **3T-T2star cervical, thoraric**    | `sct_deepseg_gm` SCT v.6.5  |
| **3T-T2star lumbar**   | `seg_gm_t2s-lumbar` , see  https://github.com/ivadomed/model-gm-contrast-region-agnostic/issues/3#issuecomment-2577525134 , [repo](https://github.com/ivadomed/lumbar-T2star-GMseg)|
| **7T-T2star**    |  `seg_gm_sc_7t_t2star` SCT v.6.5 , [repo](https://github.com/ivadomed/model_seg_gm-wm_t2star_7t_unet2d-multiclass)      | 
| **3T-PSIR**      |       `seg_gm_psir` , see : https://github.com/ivadomed/model-gm-contrast-region-agnostic/issues/3#issuecomment-2575604826                 |
| **3T-T1w**       | `seg_gm_t1w-multi-class` ,  see https://github.com/ivadomed/model-gm-contrast-region-agnostic/issues/3#issuecomment-2575865682|
| **7T-T1map**     | `seg_gm_7T-T1map` see https://github.com/ivadomed/model-gm-contrast-region-agnostic/issues/3#issuecomment-2577668214           | 
| **7T-UNIT1**     | No benchmark           |



## Conclusion
The `CRA_cropped` model works well at different contrasts and regions, @jcohenadad  do you think I can create a first release of this model on `sct_deepseg` ?
